### PR TITLE
Introduce random seed in simulate

### LINKED ISSUE #2053: Remove unused fields from `ModelCheckerParams`

### LINKED ISSUE #2054: `search.simulation` wired via tuning options

### LINKED ISSUE #2083: Add an option to seed the random seed in `simulate`
Currently, the command `simulate` is using randomization to pick the next action. To make the experiments reproducible, we should introduce an option to set the seed.

### PR DESCRIPTION
**Pull request by konnov**
Closes #2083. Closes #2054. Closes #2053. This PR adds `--seed=N` to the commands `simulate` and `check`. This seed is used by `simulate` for random selection of the next transition. In addition, both in `simulate` and `check`, if SMT seeds are not set, then `--seed=N` is also used to set the SMT seeds. Seed generation is now done at the configuration phase, if the seed is not specified.

To address #2054, we introduce `searchKind` and `maxRun` directly in the Patch/Validated pair, so the options of `simulate` are resolved by the configuration loader, instead of the late resolution in `FineTuningOptions`.

This feature is implemented with Codex GPT-5.6-sol xhigh/max and then manually adapted.

- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [x] Documentation added for any new functionality
- [x] [Entries added to `./unreleased/`][changelog format] for any new functionality

[changelog format]: https://github.com/apalache-mc/apalache/blob/main/CONTRIBUTING.md#how-to-record-a-change

### CODE REVIEW

**Review by thpani — COMMENTED** [review]
I don't understand why we need 3 different tuning options to set SMT seeds.

I think it makes more sense to make a breaking change and introduce a new global option that replaces the existing ones.

**Comment by konnov** [comment]
> I don't understand why we need 3 different tuning options to set SMT seeds.
> 
> I think it makes more sense to make a breaking change and introduce a new global option that replaces the existing ones.

I agree. We can leave one option for setting the SMT seed. The issue is that `simulate --seed=N` sets the seed for transition selection. I thought of deriving the SMT seed from this one, if no explicit SMT seed is provided. Do you think we should keep the default SMT seed, if `--seed` is provided?

**Comment by thpani** [comment]
If we have a first-class `--seed` argument, it should be the only way to set the seed, I think

**Comment by konnov** [comment]
> If we have a first-class `--seed` argument, it should be the only way to set the seed, I think

This sounds reasonable. I don't think we have users who need particular seed settings for SMT solvers.

**Comment by konnov** [comment]
OK, fixed this PR to use `--seed=N` and its internal mapping to `search.seed` as the only way to set the seeds of the transition selection and SMT.

**Comment by konnov** [comment]
I've done the further options refactoring, since we've merged #3409

File: .unreleased/breaking-changes/typed-search-controls.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Replace the tuning keys `search.seed`, `search.simulation`,
+`search.simulation.maxRun`, and `search.outputTraces` with the typed checker
```
Comment (thpani) [inline .unreleased/breaking-changes/typed-search-controls.md]: afaict, we are also retiring `smt.randomSeed`?

File: docs/src/apalache/config.md [hunk 2]
```diff
@@ -86,6 +89,10 @@ In the table below, a default of "none" means that the value is optional.
 |               | `output`             | Write the processed module to this file.                     | path string                  | none; `.tla` and `.json` are supported      |
```
Comment (thpani) [inline docs/src/apalache/config.md:92]: now that I'm reading this again, I'm not sure how the options are grouped. e.g., `source` and `output` are top-level, but `server` won't be using them?

File: docs/src/apalache/running.md [hunk 3]
```diff
@@ -79,6 +80,8 @@ The arguments are as follows:
       and integer arithmetic; `arrays` (experimental) and `funArrays` (experimental) use SMT arrays with extensionality.
       This parameter can also be set via the `SMT_ENCODING` environment variable. See the [alternative SMT encoding
       using arrays] for details.
+    - `--seed=NUM` sets a nonnegative search seed. The seed initializes SMT solving and, with `simulate`, is also used
```
Comment (thpani) [inline docs/src/apalache/running.md]: I think it would be useful to specify the datatype.
I frequently find myself wondering (for other tools) from what range I should sample seeds

File: mod-tool/src/test/scala/at/forsyte/apalache/tla/tooling/opt/TestCommandConfig.scala [hunk 4]
```diff
@@ -69,25 +69,52 @@ class TestCommandConfig extends AnyFunSuite {
...
-    assert(explicit.checker.tuning.flatMap(_.get("search.outputTraces")).contains("false"))
+    assert(explicit.checker.seed.contains(42))
+    assert(explicit.checker.outputTraces.contains(false))
+  }
+
+  test("seed is validated during resolution") {
+    val invalid = new CheckCmd()
+    invalid.read(List("CommandConfig.tla"))
+    invalid.seed = Some(-1)
+    val invalidResult = ApalacheConfigResolver.resolveCheck(invalid.toConfig.requireValue())
```
Comment (thpani) [inline mod-tool/src/test/scala/at/forsyte/apalache/tla/tooling/opt/TestCommandConfig.scala:110]: probably also worth testing overflow

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/search/ModelCheckerParams.scala [hunk 5]
```diff
@@ -34,7 +29,11 @@ object ModelCheckerParams {
 class ModelCheckerParams(
     checkerInput: CheckerInput,
     val stepsBound: Int,
-    tuningOptions: Map[String, String] = Map()) {
+    tuningOptions: Map[String, String] = Map(),
+    val searchKind: SearchKind = Check,
+    val seed: Int = 0,
```
Comment (thpani) [inline tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/search/ModelCheckerParams.scala]: perhaps my ignorance of the new options system, but why does this take a default parameter? wouldn't it more appropriate to force the caller to come op with a value? or alternatively, to accept Option[Int]?

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/search/ModelCheckerParams.scala [hunk 6]
```diff
@@ -34,7 +29,11 @@ object ModelCheckerParams {
...
+    val maxRun: Int = 100,
```
Comment (thpani) [inline tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/search/ModelCheckerParams.scala]: similar here; afaict this only applies to `simulate`? should it be `Option[Int]`?

I find it odd that the default `100` is overridden in `SearchState` for `.searchKind != Simulate`. Certainly doesn't help with debugging.

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/search/ModelCheckerParams.scala (hunk 5 above)
Comment (thpani) [reply tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/search/ModelCheckerParams.scala]: otherwise, we should say in the docs that the default seed is always fixed at 0

File: tla-io/src/main/scala/at/forsyte/apalache/io/config/ApalacheConfigResolver.scala [hunk 7]
```diff
@@ -101,7 +102,16 @@ object ApalacheConfigResolver {
     val solver = requireDefault(checkerWithDefaults.smtSolver, s"$CHECKER.$SMT_SOLVER")
     val encoding = requireDefault(checkerWithDefaults.smtEncoding, s"$CHECKER.$SMT_ENCODING")
     val maxError = requireDefault(checkerWithDefaults.maxError, s"$CHECKER.$MAX_ERROR")
+    val maxRun = requireDefault(checkerWithDefaults.maxRun, s"$CHECKER.$MAX_RUN")
 
+    configuredChecker.seed.foreach { seed =>
+      if (seed < 0) {
+        errors += s"Option $CHECKER.$SEED must be nonnegative, but got $seed."
+      }
```
Comment (thpani) [inline tla-io/src/main/scala/at/forsyte/apalache/io/config/ApalacheConfigResolver.scala:116]: what's the upper bound? I'm not even sure it's uniform between cvc5 and z3

File: tla-io/src/main/scala/at/forsyte/apalache/io/config/ApalacheConfigResolver.scala [hunk 8]
```diff
@@ -101,7 +102,16 @@ object ApalacheConfigResolver {
...
+    }
+    if (maxRun <= 0) {
+      errors += s"Option $CHECKER.$MAX_RUN must be positive, but got $maxRun."
```
Comment (thpani) [inline tla-io/src/main/scala/at/forsyte/apalache/io/config/ApalacheConfigResolver.scala:119]: should we also check `!(searchKind = simulate) => maxRun == 1`?

File: tla-io/src/main/scala/at/forsyte/apalache/io/config/ApalacheConfigResolver.scala [hunk 9]
```diff
@@ -273,6 +289,9 @@ object ApalacheConfigResolver {
   private def requireDefault[A](value: Option[A], field: String): A =
     value.getOrElse(throw new IllegalStateException(s"Missing built-in default for $field"))
 
+  /** Generate one concrete nonnegative seed for an otherwise unseeded checker run. */
+  private def generateSeed(): Int = ThreadLocalRandom.current().nextInt() & Int.MaxValue
```
Comment (thpani) [inline tla-io/src/main/scala/at/forsyte/apalache/io/config/ApalacheConfigResolver.scala:304]:
```suggestion
  /** Generate one concrete nonnegative seed for an otherwise unseeded checker run. */
  // Java `int`s are in two's complement, so sampling remains uniform
  private def generateSeed(): Int = ThreadLocalRandom.current().nextInt() & Int.MaxValue
```

File: docs/src/adr/025adr-cvc5-solver-backend.md [hunk 10]
```diff
@@ -120,9 +120,9 @@ The initial implementation should be intentionally conservative:
 2. Preserve the current cell/sort naming conventions so that existing logging, recording, and model-decoding assumptions
    remain easy to compare against Z3.
 3. Add a solver choice to configuration separately from `smtEncoding`; those are independent dimensions.
-4. Keep backend-neutral SMT controls separate from solver-specific options. For example, `smt.randomSeed`,
-   `smt.statsSec`, and SMT timeouts are concepts Apalache can define once and map to each backend. Low-level solver
-   parameters should remain namespaced, for example `z3.*` and `cvc5.*`.
+4. Keep backend-neutral SMT controls separate from solver-specific options. For example, `checker.seed`, `smt.statsSec`,
+   and SMT timeouts are concepts Apalache can define once and map to each backend. Low-level solver parameters should
+   remain namespaced, for example `z3.*` and `cvc5.*`.
```
Comment (thpani) [inline docs/src/adr/025adr-cvc5-solver-backend.md]: I'm a bit torn about modifying ADRs in this way.

This particular edit is certainly useful, but in principle we have to assume that ADRs are historical.

**Review by thpani — COMMENTED** [review]
Looks mostly good, but I think we need another round on this

File: docs/src/adr/025adr-cvc5-solver-backend.md (hunk 10 above)
Comment (konnov) [reply docs/src/adr/025adr-cvc5-solver-backend.md]: True. If we want to be really strict, we can introduce a revision history like in standards. Though it looks like an overkill in this case.

File: docs/src/apalache/config.md (hunk 2 above)
Comment (konnov) [reply docs/src/apalache/config.md:92]: I realized that `source` and `output` are a real security risk in the server mode. Not that Apalache itself and its server APIs are really secure, but letting an arbitrary server request to read and write arbitrary files is just asking for trouble. I am not even sure about why we need `output`. In the case of CLI, `source` is unavoidable.

Looking at the configuration options, too, yeah, we could regroup it even further.

Comment (konnov) [reply docs/src/apalache/config.md:92]: Also, `smtprof` and `profiling` are top-level, because we have `check` and `tracee` (the last one I would rather evict and make an external tool). I guess, the logic is whatever is shared between different commands goes to top-level. Previously, it was `common`, which was also confusing.

File: .unreleased/breaking-changes/typed-search-controls.md (hunk 1 above)
Comment (konnov) [reply .unreleased/breaking-changes/typed-search-controls.md]: Added to the notes

File: docs/src/apalache/running.md (hunk 3 above)
Comment (konnov) [reply docs/src/apalache/running.md]: Fixed:

```md
    - `--seed=NUM` sets a search seed in the inclusive range `0..2147483647`. The seed initializes SMT solving and, with
      `simulate`, is also used for transition selection. When omitted, Apalache generates and logs a fresh seed for the
      run.
```

File: mod-tool/src/test/scala/at/forsyte/apalache/tla/tooling/opt/TestCommandConfig.scala (hunk 4 above)
Comment (konnov) [reply mod-tool/src/test/scala/at/forsyte/apalache/tla/tooling/opt/TestCommandConfig.scala:110]: done:

```scala
  test("seed parsing accepts the upper bound and rejects overflow") {
    val boundary = new CheckCmd()
    boundary.read(List(s"--$SEED=${Int.MaxValue}", "CommandConfig.tla"))
    val config = ApalacheConfigResolver.resolveCheck(boundary.toConfig.requireValue()).requireValue()
    assert(config.checker.seed == Int.MaxValue)

    val overflow = intercept[ParsingException] {
      new CheckCmd().read(List(s"--$SEED=2147483648", "CommandConfig.tla"))
    }
    assert(overflow.getMessage.contains("expected an Int"))
  }
```

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/search/ModelCheckerParams.scala (hunk 5 above)
Comment (konnov) [reply tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/search/ModelCheckerParams.scala]: Good point. Now the parameters are passed directly, no defaults anymore. Also removed the unused constants that were mentioned in #2053

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/search/ModelCheckerParams.scala (hunk 6 above)
Comment (konnov) [reply tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/search/ModelCheckerParams.scala]: No defaults in `ModelCheckerParams` anymore

File: tla-io/src/main/scala/at/forsyte/apalache/io/config/ApalacheConfigResolver.scala (hunk 9 above)
Comment (konnov) [reply tla-io/src/main/scala/at/forsyte/apalache/io/config/ApalacheConfigResolver.scala:304]: Added it directly

File: tla-io/src/main/scala/at/forsyte/apalache/io/config/ApalacheConfigResolver.scala (hunk 8 above)
Comment (konnov) [reply tla-io/src/main/scala/at/forsyte/apalache/io/config/ApalacheConfigResolver.scala:119]: yep, added that

File: tla-io/src/main/scala/at/forsyte/apalache/io/config/ApalacheConfigResolver.scala (hunk 7 above)
Comment (konnov) [reply tla-io/src/main/scala/at/forsyte/apalache/io/config/ApalacheConfigResolver.scala:116]: Added the bound in the error message. We don't have to check it in the code though, as `seed` is of type `Int`.

File: docs/src/adr/025adr-cvc5-solver-backend.md (hunk 10 above)
Comment (konnov) [reply docs/src/adr/025adr-cvc5-solver-backend.md]: Added a revision note (actually, Codex did that)

**Comment by konnov** [comment]
@thpani I think I've fixed all of your comments (except the configuration layout still needs some thinking). Can you have another look?

**update:** when this is done, I want to have another look at #3419
